Fix max.poll.interval.ms setting of KafkaConsumer.

markusthoemmes opened the pull request:
The `max.poll.interval.ms` setting needs to be high enough to not be thrown off by all actions in the system consuming the maximum duration of their respective action timeout, ideally also adding enough slack to account for image pulls taking long, or docker runs being very slow.

The name of the setting was also misspelled so we've been running on the default.


<!--- Provide a concise summary of your changes in the Title -->

## Description
<!--- Provide a detailed description of your changes. -->
<!--- Include details of what problem you are solving and how your changes are tested. -->

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [ ] Controller
- [X] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [X] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [X] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).

chetanmeh commented on common/scala/src/main/resources/application.conf:
```diff
@@ -72,7 +72,7 @@ whisk {
             enable-auto-commit = false
             auto-offset-reset = earliest
-            max-poll-interval = 360000
+            max-poll-interval-ms = 900000 // 20 minutes
```
May be change it to use duration as a unit. `max-poll-interval = 20 minutes`

markusthoemmes replied:
These values are passed directly down to the kafka-setting which don't support durations per my knowledge.

The configuration here is build to be passed directly to the respective consumer/producer without touching it in our code at all.

chetanmeh replied:
Did not knew that. So configs as mentioned [here][1] are passed as is. May be we can later add a testcase which uses [AdminClient.describeTopic][2] to assert the expected config

[1]: https://kafka.apache.org/documentation/#newconsumerconfigs
[2]: http://kafka.apache.org/0110/javadoc/index.html?org/apache/kafka/clients/admin/AdminClient.html

sven-lange-last replied:
Very good finding! And: 900,000 ms are 15 minutes :-)

sven-lange-last commented:
The Kafka new Java consumer documentation contains the following hint in the [notables changes section for 0.10.1.0](https://kafka.apache.org/documentation/#upgrade_1010_notable):

```
The new Java Consumer now supports heartbeating from a background thread. There is a new configuration max.poll.interval.ms which controls the maximum time between poll invocations before the consumer will proactively leave the group (5 minutes by default). The value of the configuration request.timeout.ms must always be larger than max.poll.interval.ms because this is the maximum time that a JoinGroup request can block on the server while the consumer is rebalancing, so we have changed its default value to just above 5 minutes.
```

At the moment, we have `request.timeout.ms` < `max.poll.interval.ms` and this PR will also keep the value smaller.

I guess that this should not harm in our situation with a single consumer group per topic, a single partition per topic and a single consumer process in the consumer group. But I don't understand all implications...

markusthoemmes commented:
Fair point @sven-lange-last, added the relevant timeout + doc link.

@chetanmeh I added a check at startup of either Consumer/Producer that verifies the config and prints an error if all set keys are known.

sven-lange-last commented:
From [Kafka documentation - New Consumer Configs](https://kafka.apache.org/documentation/#newconsumerconfigs):

* `request.timeout.ms`: The configuration controls the maximum amount of time the client will wait for the response of a request. If the response is not received before the timeout elapses the client will resend the request if necessary or fail the request if retries are exhausted.
* `max.poll.interval.ms`: The maximum delay between invocations of poll() when using consumer group management. This places an upper bound on the amount of time that the consumer can be idle before fetching more records. If poll() is not called before expiration of this timeout, then the consumer is considered failed and the group will rebalance in order to reassign the partitions to another member.

chetanmeh commented on common/scala/src/main/scala/whisk/connector/kafka/KafkaMessagingProvider.scala:
```diff
@@ -101,4 +101,24 @@ object KafkaConfiguration {
+    val unknownSettings = config -- knownKeys
+
+    if (unknownSettings.nonEmpty) {
+      logging.warn(this, s"potential misconfiguration, unknown settings: ${unknownSettings.mkString(",")}")
```
May be just log key names as value may potentially contain sensitive info

sven-lange-last replied:
Sounds reasonable.
